### PR TITLE
[Merged by Bors] - syncer: separate data syncing and processing

### LINKED ISSUE #3164: sync: data sync keeps going while validation stuck
## Description

at the beginning of every epoch, there is a period of time when beacon is unavailable (until enough ballots are seen by the node and beacon value determined)
during this time, syncer will see that beacon is not available and will not request tortoise to verify a layer.

but the problem is: syncer uses two goroutines
goroutine 0: sync data from processedLayer (say 100) -> currentLayer-1 (say 10_000)
goroutine 1: request tortoise to verify layer

say at the beginning of epoch 3 (layer 120) the above beacon check stops goroutine1 but not goroutine0. so syncer will end up syncing data all the way from 120 -> 10_000 (while verified layer get stuck at 119) for goroutine0 to exit, and allow the next syncer run to proceed.

this is wildly inefficient.

## Steps to reproduce
1. starts a new node and one can see log message from syncer `skip validating layer: beacon not available`

## Actual Behavior

see above

## Expected Behavior

verifying a layer should be able to progress after the beacon value is available for that epoch.

### PR DESCRIPTION
**Pull request by countvonzero**
## Motivation

Closes #3164
Part of #2415

## Changes

separate the ballots processing of each layer into a separate long-running go routine that keeps trying from mesh.LatestLayerInState() to syncer's last synced layer

## Test Plan

unit test, systest, manual testing

## DevOps Notes

- [x] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [x] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [x] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

### CODE REVIEW

**Comment by countvonzero** [comment]
bors try

**Comment by countvonzero** [comment]
bors try

**Comment by countvonzero** [comment]
> ## try
> Build failed:
> 
> * [ci-stage2](https://github.com/spacemeshos/go-spacemesh/runs/6380256963?check_suite_focus=true)

@dshulyak is this a legit failure?

```
=== CONT  TestAddNodes
    nodes_test.go:90: 
        	Error Trace:	nodes_test.go:90
        	Error:      	"map[0x02f5c02372aad79a6fa0dbe934b8a3a8a4dee5a8f96e4ed[67](https://github.com/spacemeshos/go-spacemesh/runs/6379893565?check_suite_focus=true#step:11:67)343e1f1e2eae6ba:{} 0x1353c7e22358b4e1970d6f461d35bd2de3950943f0ad889887f33d5e76e01466:{} 0x172c10a575e521d12d2d36005057cc49473d9f16fc4f8d245e008a28b5761447:{} 0x2bc877f6c0c9bf49d86c38[68](https://github.com/spacemeshos/go-spacemesh/runs/6379893565?check_suite_focus=true#step:11:68)89789c7d1428ef201cc6be05196b6c903b7ffb13:{} 0x32b29668a4b4f0d66d523ff7f88c837cd0f416ebc1b5e49dc4c5efe6eec8f7e9:{} 0x35662c67fb26acd1ae198b46091df7d7c336489806d375be05e08585f3ce2da1:{} 0x389a600482232e451912dc7e35d2de81d209beca566761464af35f2ab9645d34:{} 0x3b2fa90f7df33257baf2c3fa5fcb83ff08fdba1d9e75df22430bd3fd8563eb72:{} 0x3d4c1831529428134d71c2f819d7f6d10967bc8815d9fd316bc7287d272e8537:{} 0x4616e8fd2edda0cf3d142492418dad78567f3ceff5608809b29ef78a2c24433c:{} 0x605e3b50b3124a5884867714c7e3fca9fdf41f9e2bf9cd35dcbb7ef388784059:{} 0x[69](https://github.com/spacemeshos/go-spacemesh/runs/6379893565?check_suite_focus=true#step:11:69)60165f5542c3a8d34c6456b393c65316e6a07630844fd15782d54728b13d38:{} 0x7526d8bc7ddccb024fc35334c26d0f5ce82dab040c6f597b7[70](https://github.com/spacemeshos/go-spacemesh/runs/6379893565?check_suite_focus=true#step:11:70)edba1b8c82221:{} 0x831abba1d4149f5e16338862fb00a0373833296e728abdec1304b3470823423a:{} 0x8a31899506ae88811c449122033e9ba361a086423c3d53f96aa7f58a39b4afde:{} 0x93b87545d2cfa89d4bf9823f3dbeaf8ffb86a4732c5d52eecc27decea0feaa0e:{} 0xa22703b65b9b9f8c28f7d5e11a877322869be70ac89b82a32e15e64167c4d598:{} 0xa5ff88a7545ca89729bca22c5697177f7074837d937a38e7b92d682179836bf3:{} 0xa780d486413f53182639e9753cdfb7090bcae28ed7a07c41fbde87f3a5e3750e:{} 0xb1ee43b9e9672616bf68818e5ddf1e6ee7a05e28710b0f6b84f17632060df576:{} 0xb4bdee1c4c0e771345fcd7331cc6f7a449f794303b8dc6288019235fd8c33ef5:{} 0xb52cfcb1bd057675b649eb98c9d2fbce672c39d41153597a1a4c98153ff7e41e:{} 0xbb4f42432a5a8a318a9b31483ae2a235cba807083949481368eaaa37c52333dc:{} 0xc362d08e913cc82acc46252e0286f53fdbbdc198226fa2697517d8727ca87e3b:{} 0xc5bb70d822b23d6d44882588eac9d0a5b7ba4b1bede47aeba16bf9631d563587:{} 0xc90cbc95cb72ba160d27f41c6ff5d2f7cd7e16dde3923c193d3fcaed4b658cab:{} 0xd5f47ff1203f65cb73b247cc16fe5c3b4f38e68cde4d5e01df26efb4626092b0:{} 0xd7da7df9774d866fbbe2fdc6e9313034c330da0a33bd8ff3097505ae19fc6c84:{} 0xe10b1ebad570910a103e5c946f4b8fbaf6fe87652260d26ba0e81386429c3c07:{} 0xe5b9257e0e1d76880ca45536191b33c301cc203f1477f5f5ecd9ee2279374d96:{}]" 
should have 28 item(s), but has 30
        	Test:       	TestAddNodes
        	Messages:   	epoch=5
```

**Comment by dshulyak** [comment]
@countvonzero that failure means that it was expected that sync will take more time

**Comment by countvonzero** [comment]
> @countvonzero that failure means that it was expected that sync will take more time

thanks. i relaxed it to greater or equal to

**Comment by countvonzero** [comment]
bors try

**Comment by countvonzero** [comment]
bors try

**Comment by countvonzero** [comment]
> ## try
> Build failed:
> 
> * [ci-stage2](https://github.com/spacemeshos/go-spacemesh/runs/6394313189?check_suite_focus=true)

this one failed with TestShortTimeskew
```
=== CONT  TestShortTimeskew
    assertion_compare.go:342: 
        	Error Trace:	timeskew_test.go:66
        	Error:      	"11" is not less than or equal to "9"
        	Test:       	TestShortTimeskew
        	Messages:   	[]
    logger.go:130: 2022-05-11T19:05:44.374Z	INFO	namespace was deleted	{"namespace": "test-hghy"}
--- FAIL: TestShortTimeskew (1914.94s)
FAIL
```

**Comment by countvonzero** [comment]
bors try

**Comment by countvonzero** [comment]
bors try

**Comment by countvonzero** [comment]
bors try

File: syncer/syncer.go [hunk 1]
```diff
@@ -299,91 +303,63 @@ func (s *Syncer) synchronize(ctx context.Context) bool {
...
-		// number of attempts to start a validation goroutine
-		attempt = 0
-		// whether the data sync succeed. validation failure is not checked by design.
-		success = true
-		// the last layer that is synced in this run
-		lastSynced = s.mesh.ProcessedLayer()
-	)
-
-	attemptFunc := func() bool {
+	syncFunc := func() bool {
```
Comment (evt) [inline syncer/syncer.go:318]: why do we not define it as a static function?

File: syncer/syncer.go [hunk 2]
```diff
@@ -50,9 +28,6 @@ const (
 
 	// the max amount of layer delays syncer can tolerate before it tries to validate a layer.
 	maxHareDelayLayers uint32 = 10
-
-	// the max number of attempts syncer tries to advance the processed layer within a sync run.
-	maxAttemptWithinRun = 3
```
Comment (evt) [inline syncer/syncer.go:55]: do we not need attempts anymore?

Comment (countvonzero) [reply syncer/syncer.go:55]: the reason why we used to have multiple attempts is that we would sync a layer, and then it fails when mesh fails to process the layer. so we would let it try max attempts times to process after we sync the data. now that the two (syncing data and processing ballots) are separated into two goroutines, there is no such need anymore

File: syncer/syncer.go (hunk 1 above)
Comment (countvonzero) [reply syncer/syncer.go:318]: this function is not strictly static. it needs struct member (mesh/ticker) and method (s.getLastSyncedLayer()/s.setLastSyncedLayer()). i don't think the code would be cleaner changing this to a static func.

**Review by evt — APPROVED** [review]

**Comment by countvonzero** [comment]
bors merge

**Pull request closed without merging**